scm-plugins/scm-hg-plugin/src/main/java/sonia/scm/repository/hooks/HookServer.java

```diff
@@ -0,0 +1,143 @@
+ * SOFTWARE.
+ */
+
+package sonia.scm.repository.hooks;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.shiro.SecurityUtils;
+import org.apache.shiro.mgt.SecurityManager;
+import org.apache.shiro.util.ThreadContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+@Singleton
+public class HookServer implements AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HookServer.class);
+
+  private final HookHandlerFactory handlerFactory;
+
+  private ExecutorService acceptor;
+  private ExecutorService workerPool;
+  private ServerSocket serverSocket;
+
+  @Inject
+  public HookServer(HookHandlerFactory handlerFactory) {
+    this.handlerFactory = handlerFactory;
+  }
+
+  private ExecutorService createAcceptor() {
+    return Executors.newSingleThreadExecutor(
+      createThreadFactory("HgHookAcceptor")
+    );
+  }
+
+  private ExecutorService createWorkerPool() {
+    return Executors.newCachedThreadPool(
+      createThreadFactory("HgHookWorker-%d")
+    );
+  }
+
+  @Nonnull
+  private ThreadFactory createThreadFactory(String hgHookAcceptor) {
+    return new ThreadFactoryBuilder()
+      .setNameFormat(hgHookAcceptor)
+      .build();
+  }
+
+  public int start() throws IOException {
+    acceptor = createAcceptor();
+    workerPool = createWorkerPool();
+    serverSocket = createServerSocket();
+    // set timeout to 2 min, to avoid blocking clients
+    serverSocket.setSoTimeout(2 * 60 * 1000);
+
+    accept();
+
+    int port = serverSocket.getLocalPort();
+    LOG.info("open hg hook server on port {}", port);
+    return port;
+  }
+
+  private void accept() {
+    SecurityManager securityManager = SecurityUtils.getSecurityManager();
+    acceptor.submit(() -> {
+      while (!serverSocket.isClosed()) {
+        try {
+          LOG.trace("wait for next hook connection");
+          Socket clientSocket = serverSocket.accept();
+          LOG.trace("accept incoming hook client from {}", clientSocket.getInetAddress());
+          HookHandler hookHandler = handlerFactory.create(clientSocket);
+          workerPool.submit(associateSecurityManager(securityManager, hookHandler));
+        } catch (IOException ex) {
+          LOG.debug("failed to accept socket, possible closed", ex);
+        }
+      }
+      LOG.warn("ServerSocket is closed");
+    });
+  }
+
+  private Runnable associateSecurityManager(SecurityManager securityManager, HookHandler hookHandler) {
+    return () -> {
+      ThreadContext.bind(securityManager);
+      try {
+        hookHandler.run();
+      } finally {
+        ThreadContext.unbindSubject();
```

Comment:
Why would you unbind the subject here?

Reply:
This because the hook handler does a login, which binds a subject to the thread. We use a cached thread pool, which reuse existing threads and to avoid that a subject is used from a prior hook, we have to unbind it.